src/boot/fwupd.c

```diff
@@ -0,0 +1,589 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+/* Ported from:
+ * https://github.com/fwupd/fwupd-efi/blob/1a794fc7ce481176926687f4e88e6e0c8d5e9699/efi/fwupdate.c
+ * Copyright (C) 2014 Red Hat, Inc.
+ */
+
+#include "device-path-util.h"
+#include "efi.h"
+#include "efi-efivars.h"
+#include "efi-log.h"
+#include "fwupd.h"
+#include "string-util-fundamental.h"
+#include "util.h"
+
+#define FWUPDATE_ATTEMPT_UPDATE 0x00000001
+#define FWUPDATE_ATTEMPTED      0x00000002
+
+#define CAPSULE_FLAGS_PERSIST_ACROSS_RESET    0x00010000
+#define CAPSULE_FLAGS_POPULATE_SYSTEM_TABLE   0x00020000
+#define CAPSULE_FLAGS_INITIATE_RESET          0x00040000
+
+#define EFI_GRAPHICS_OUTPUT_PROTOCOL_GUID { 0x9042a9de, 0x23dc, 0x4a38, {0x96, 0xfb, 0x7a, 0xde, 0xd0, 0x80, 0x51, 0x6a } }
+#define UX_CAPSULE_GUID {0x3b8c8162, 0x188c, 0x46a4, {0xae, 0xc9, 0xbe, 0x43, 0xf1, 0xd6, 0x56, 0x97} }
+
+typedef struct {
+        uint32_t max_mode;
+        uint32_t mode;
+        /*EFI_GRAPHICS_OUTPUT_MODE_INFORMATION*/ void *info;
+        uint64_t size_of_info;
+        EFI_PHYSICAL_ADDRESS frame_buffer_base;
+        uint64_t frame_buffer_size;
+} EFI_GRAPHICS_OUTPUT_PROTOCOL_MODE;
+
+typedef struct  {
+        void *query_mode;
+        void *set_mode;
+        void *blt;
+        EFI_GRAPHICS_OUTPUT_PROTOCOL_MODE *mode;
+} EFI_GRAPHICS_OUTPUT_PROTOCOL;
+
+typedef struct {
+        uint8_t version;
+        uint8_t checksum;
+        uint8_t image_type;
+        uint8_t reserved;
+        uint32_t mode;
+        uint32_t x_offset;
+        uint32_t y_offset;
+} _packed_ UX_CAPSULE_HEADER;
```

Comment:
hmm, where's this structure from? some efi spec? if so, it should probably not be here.

or fwupd? then plesae add coment